### PR TITLE
Object store/ rest api

### PR DESCRIPTION
**Pull request by deepsidhu85**
## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

Updated REST API to support object storage

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

~* [ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.~
~* [ ] Tests added (or description of how to test) for any new features.~
~* [ ] User documentation updated for UI or technical changes.~

### CODE REVIEW

**Review by tom114 — CHANGES_REQUESTED** [review]
Thanks for the work here @deepsidhu85 .  A few questions and modifications below.

File: src/main/java/ca/corefacility/bioinformatics/irida/model/assembly/GenomeAssembly.java [hunk 1]
```diff
@@ -154,4 +154,8 @@ public boolean equals(Object obj) {
 	public InputStream getFileInputStream() {
 		return IridaFiles.getFileInputStream(getFile());
 	}
+
+	public String getStorageType(){
+		return IridaFiles.getStorageType();
+	}
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/assembly/GenomeAssembly.java:160]: When using the rest api to list a sample and its files it shows you where the file is located. Currently only displays the location depending on the value set in the irida conf file for the storage type

File: src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/SequenceFile.java [hunk 2]
```diff
@@ -309,4 +309,17 @@ public InputStream getFileInputStream() {
...
+	 *
+	 * @return if file exists or not
+	 */
+	public Long getFileSizeBytes() {
+		return IridaFiles.getFileSizeBytes(getFile());
+	}
+
+	public String getStorageType(){
+		return IridaFiles.getStorageType();
+	}
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/SequenceFile.java:323]: Same as above

File: src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/SequenceFile.java [hunk 3]
```diff
@@ -309,4 +309,17 @@ public InputStream getFileInputStream() {
 		return IridaFiles.getFileInputStream(file);
 	}
 
+	/**
+	 * Get file size in bytes
+	 *
+	 * @return if file exists or not
+	 */
+	public Long getFileSizeBytes() {
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/SequenceFile.java:317]: This method is still used when displaying the size of files in a sample and sequencing run. I will end up updating this during my sample details refactor so just going to leave it for now

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAwsUtilityImpl.java [hunk 4]
```diff
@@ -49,6 +51,7 @@ public IridaFileStorageAwsUtilityImpl(String bucketName, String bucketRegion, St
 				.withCredentials(new AWSStaticCredentialsProvider(awsCreds))
 				.build();
 		this.bucketName = bucketName;
+		this.storageType = StorageType.AWS;
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAwsUtilityImpl.java]: Updated in a256989b5b322f9e57d2818a7127f21d99ea05d3

File: src/main/java/ca/corefacility/bioinformatics/irida/web/spring/view/CSVView.java [hunk 5]
```diff
@@ -46,10 +49,15 @@ protected void renderMergedOutputModel(Map<String, Object> model, HttpServletReq
...
 		response.setHeader(HttpHeaders.CONTENT_TYPE, DEFAULT_CONTENT_TYPE);
-		response.setHeader(HttpHeaders.CONTENT_LENGTH, String.valueOf(Files.size(fileContent)));
-		OutputStream os = response.getOutputStream();
-		Files.copy(fileContent, os);
-		os.flush();
-		os.close();
+		response.setHeader(HttpHeaders.CONTENT_LENGTH, String.valueOf(sfr.getFileSizeBytes()));
+
+		try(InputStream is = sfr.getFileInputStream()) {
+			OutputStream os = response.getOutputStream();
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/spring/view/CSVView.java]: Updated in a256989b5b322f9e57d2818a7127f21d99ea05d3

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/RESTAnalysisSubmissionController.java [hunk 6]
```diff
@@ -313,6 +318,28 @@ public RESTAnalysisSubmissionController(AnalysisSubmissionService analysisSubmis
...
+	 * @throws IOException if the file input stream cannot be read
+	 */
+	@Operation(operationId = "getAnalysisOutputFileContents", summary = "Get the analysis output file contents", description = "Get the analysis output file contents for the given analysis submission.", tags = "analysisSubmissions")
+	@RequestMapping(value = "/{submissionId}/analysis/file/{fileId}", produces = MediaType.TEXT_PLAIN_VALUE)
+	@ResponseBody
+	public ResponseResource<ByteArrayResource> getAnalysisOutputFileContents(@PathVariable Long submissionId,
+			@PathVariable Long fileId) throws IOException {
+		AnalysisOutputFile analysisOutputFile = getOutputFileForSubmission(submissionId, fileId);
+		try (InputStream inputStream = analysisOutputFile.getFileInputStream()) {
+			ByteArrayResource fileContents = new ByteArrayResource(inputStream.readAllBytes());
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/RESTAnalysisSubmissionController.java]: This method is being re-added into the rest api in a separate PR. I will update it to not read the entire output file into memory once that PR is merged up

File: src/main/java/ca/corefacility/bioinformatics/irida/model/assembly/GenomeAssembly.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/assembly/GenomeAssembly.java:159]: Ok thanks.  This is likely a good addition.

File: src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/SequenceFile.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/SequenceFile.java:318]: Thanks I see this is being ignored now.  It's an expensive call if we have it being run in the REST API when it isn't being used.  If you're going to remove one of the old methods, can you throw a `@Deprecated` on the one you're planning to remove?  That way we can mark which one is going to go away.

[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/SequenceFile.java:318]: Sure! Will add that

[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/model/sequenceFile/SequenceFile.java:318]: My mistake. Looking into it further both the getFileSizeBytes and getFileSize methods are used. There are certain places where we require a human readable file size in the UI and certain places where we only require the size in bytes

**Comment by deepsidhu85** [comment]
@ericenns Here is a screenshot of the rest api call that shows the representation of the file and its storage location:

restapicall

**Review by ericenns — APPROVED** [review]
I think this looks good enough to merge. Not sure exactly how I feel about returning file path in the rest api for non local storage files, but I understand why you left it. It could still be addressed in the larger PR.

**Pull request merged**